packages/medusa/src/utils/authenticate-middleware.ts

```diff
@@ -32,50 +35,40 @@ export const authenticate = (
   ): Promise<void> => {
     const authTypes = Array.isArray(authType) ? authType : [authType]
 
-    // @ts-ignore
-    const session: MedusaSession = req.session || {}
+    // We only allow authenticating using a secret API key on the admin
+    if (authTypes.includes(API_KEY_AUTH) && isAdminScope(authScope)) {
+      const apiKey = await getApiKeyInfo(req)
+      if (apiKey) {
+        ;(req as AuthenticatedMedusaRequest).auth = {
+          actor_id: apiKey.id,
+          // TODO: Should we set this to the creator of the token (apiKey.created_by), or do we leave it empty?
+          auth_user_id: "",
+          app_metadata: {},
+          // TODO: Add more limited scope once we have support for it in the API key module
+          scope: "admin",
+        }
 
-    let authUser: AuthUserDTO | null = null
-    if (authTypes.includes(SESSION_AUTH)) {
-      if (
-        session.auth_user &&
-        stringEqualsOrRegexMatch(authScope, session.auth_user.scope)
-      ) {
-        authUser = session.auth_user
+        return next()
```

Comment:
question: Having `return next()` here means you cannot authenticate as a user on requests that carry a bearer token. Problably an edge case, but do we want to allow for that?

I would have thought just passing through all the configured auth types would make sense. Obviously, returning early if `req.auth` is set in one of them.

Reply:
@olivermrbl This was intentional, but whether we want it like that or not I am happy to discuss 😄 

My thought-process was that we need to have some ordering/precendence when doing authentication. I thought an API key to be the most specialized case, and as such, should be the first on the list. If you haven't authenticated with an API key, then we fall back to other modes of authentication.

I think there is no right or wrong solution here, I personally just like not having too many options as it adds confusion to both the implementation side, as well as the user side. Let me know what you think though, I am not set on this approach :)

Reply:
> I thought an API key to be the most specialized case, and as such, should be the first on the list

Agree with the precedence and your priority. 

> then we fall back to other modes of authentication.

This is what I want to allow, but with the current implementation, I don't think this is possible, because we return `next()` regardless of the result of the API key auth no? In case API key authentication fails, we should proceed to try the other types of authentication.

One scenario I can think of where returning `next()` early could be an issue would be if you've configured you have configured the SDK in your client with an API key (and this client has regular user sign in too), and you choose to revoke that API key, then all your users would get 401s on attempted sign-in. Only when you have removed or updated the API key will your users be able to sign in again.

As I said, this is probably an edge case since you are unlikely to use the bearer token, if you also have regular user session sign-in.

Reply:
@olivermrbl If you see, we only call `return next()` only if the `if(apiKey)` is true, and that is only true if the API key is valid, so it will fallback to other modes of authentication, but I'll add it in the test scenario.

Reply:
Oh man, I'm so sorry for going back and forth on this. Missed a curly bracket.

But I am glad that we agree 😄

Reply:
Added one more test where we try to use a revoked key

Reply:
@olivermrbl no worries, it could also have been a bug so thanks for pushing 😄